.cbmc-batch/source/openssl/evp_override.c

```diff
@@ -0,0 +1,347 @@
+        if (pkey->references == 0) {
+            EC_KEY_free(pkey->ec_key);
+            free(pkey);
+        }
+    }
+}
+
+enum evp_aes { EVP_AES_128_GCM, EVP_AES_192_GCM, EVP_AES_256_GCM };
+
+/* Abstraction of the EVP_CIPHER struct */
+struct evp_cipher_st {
+    enum evp_aes from;
+};
+
+/*
+ * Description: AES for 128, 192 and 256 bit keys in Galois Counter Mode (GCM). These ciphers require additional control
+ * operations to function correctly, see the "AEAD Interface" in EVP_EncryptInit(3) section for details. Return values:
+ * These functions return an EVP_CIPHER structure that contains the implementation of the symmetric cipher.
+ */
+const EVP_CIPHER *EVP_aes_128_gcm(void) {
+    static const EVP_CIPHER cipher = { EVP_AES_128_GCM };
+    return &cipher;
+}
+const EVP_CIPHER *EVP_aes_192_gcm(void) {
+    static const EVP_CIPHER cipher = { EVP_AES_192_GCM };
+    return &cipher;
+}
+const EVP_CIPHER *EVP_aes_256_gcm(void) {
+    static const EVP_CIPHER cipher = { EVP_AES_256_GCM };
+    return &cipher;
+}
+
+enum evp_sha { EVP_SHA256, EVP_SHA384, EVP_SHA512 };
+
+/* Abstraction of the EVP_MD struct */
+struct evp_md_st {
+    enum evp_sha from;
+    size_t size;
+};
+
+/*
+ * Description: The SHA-2 SHA-224, SHA-256, SHA-512/224, SHA512/256, SHA-384 and SHA-512 algorithms, which generate 224,
+ * 256, 224, 256, 384 and 512 bits respectively of output from a given input. Return values: These functions return a
+ * EVP_MD structure that contains the implementation of the symmetric cipher.
+ */
+const EVP_MD *EVP_sha256() {
+    static const EVP_MD md = { EVP_SHA256, 32 };
+    return &md;
+}
+const EVP_MD *EVP_sha384() {
+    static const EVP_MD md = { EVP_SHA384, 48 };
+    return &md;
+}
+const EVP_MD *EVP_sha512() {
+    static const EVP_MD md = { EVP_SHA512, 64 };
+    return &md;
+}
+
+/* Abstraction of the EVP_MD_CTX struct */
+struct evp_md_ctx_st {
+    bool is_initialized;
+    EVP_PKEY *pkey;
+    size_t digest_size;
+};
+
+/*
+ * Description: Allocates and returns a digest context.
+ */
+EVP_MD_CTX *EVP_MD_CTX_new() {
+    EVP_MD_CTX *ctx = can_fail_malloc(sizeof(EVP_MD_CTX));
+
+    if (ctx) {
+        ctx->is_initialized = false;
+        ctx->pkey           = NULL;
+        ctx->digest_size    = 0;
+    }
+
+    return ctx;
+}
+
+/*
+ * Description: Cleans up digest context ctx and frees up the space allocated to it.
+ */
+void EVP_MD_CTX_free(EVP_MD_CTX *ctx) {
+    if (ctx) {
+        EVP_PKEY_free(ctx->pkey);
+        free(ctx);
+    }
+}
+
+/*
+ * Description: Sets up digest context ctx to use a digest type from ENGINE impl. type will typically be supplied by
+ * a function such as EVP_sha1(). If impl is NULL then the default implementation of digest type is used. Return
+ * values: Returns 1 for success and 0 for failure.
+ */
+int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl) {
+    assert(ctx);
+    assert(!ctx->is_initialized);
+    assert(evp_md_is_valid(type));
```

Comment:
Does this include type is nonnull?  I find it interesting that elsewhere you do seem to separate out nonnull from the valid predicate.  My valid predicates always include nonnull.

Reply:
It does include that it is non-null.